doc/specs/#885 - Terminal Settings Model/Actions Addendum.md

```diff
@@ -0,0 +1,310 @@
+    - See [Unbinding actions](#unbinding-actions) in potential issues.
+
+
+## UI/UX Design
+
+N/A
+
+## Capabilities
+
+N/A
+
+### Accessibility
+
+N/A
+
+### Security
+
+N/A
+
+### Reliability
+
+N/A
+
+### Compatibility
+
+N/A
+
+### Performance, Power, and Efficiency
+
+## Potential Issues
+
+### Layering Actions
+
+We need a way to determine where an action came from to minimize how many actions we serialize when we
+ write to disk. This is a two part approach that happens as we're loading the settings
+ 1. Load defaults.json
+     - For each of the actions in the JSON...
+         - Construct the `Command` (basically the `Command::LayerJson` we have today)
+         - Add it to the `ActionMap`
+           - this should update the internal state of `ActionMap` appropriately
+           - if the newly added key chord conflicts with a pre-existing one,
+              redirect `_KeyMap` to the newly added `Command` instead,
+              and update the conflicting one.
+2. Load settings.json
+     - Create a child for the `ActionMap`
+         - The purpose of a parent is to continue a search when the current `ActionMap` can't find a `Command` for a query. The parent is intended to be immutable.
+     - Load the actions array like normal into the child (see step 1)
+
+Introducing a parent mechanism to `ActionMap` allows it to understand where a `Command` came from. This allows us to minimize the number of actions we serialize when we write to disk, as opposed to serializing the entire list of actions.
+
+`ActionMap` queries will need to check their parent when they cannot find a matching `InternalActionID` in their `_ActionMap`.
+
+Since `NameMap` is generated upon request, we will need to pass a `std::set<InternalActionID>` as we generate the `NameMap` across each layer. This will ensure that each `Command` is only added to the `NameMap` once. We will start from the current layer and move up the inheritance tree to ensure that the current layer takes priority.
+
+Nested commands will be saved to their own map since they do not have an `InternalActionID`. To ensure layering is accomplished correctly, we will need to start from the top-most parent and update `NameMap`. As we go down the inheritance tree, any conflicts are resolved by prioritizing the current layer (the child). This ensures that the current layer takes priority. 
+
+### Modifying Actions
+
+There are several ways a command can be modified:
+- change/remove the key chord
+- change the name
+- change the icon
+- change the action
+
+It is important that these modifications are done through `ActionMap` instead of `Command`.
+ This is to ensure that the `ActionMap` is always aligned with `Command`'s values. `Command` should only expose getters in the projected type to enforce this. Thus, we can add the following functions to `ActionMap`:
+
+ ```c++
+ runtimeclass ActionMap

```

Comment:
should we have an interface for `IActionMapView` so that we can pass around read-only personas of the ActionMap?

Reply:
That's a good idea. TermApp honestly has no reason to ever call these setters. So we can just expose `IActionMapView` to TermApp and `ActionMap` to TerminalSettingsEditor.